tekton/utils.go

```diff
@@ -107,11 +107,19 @@ func hasPipelineRunStateChangedToDeleting(objectOld, objectNew client.Object) bo
 	return false
 }
 
-// isPipelineRunSigned returns a boolean indicated whether the PipelineRun been signed
-// If the object passed to this function is not a PipelineRun, the function will return false.
-func isPipelineRunSigned(objectNew client.Object) bool {
+// isChainsDoneWithPipelineRun returns a boolean indicating whether Tekton Chains is done processing
+// the PipelineRun. true is returned regardless if Chains was able to successfully sign/attest the
+// artifacts produced by the PipelineRun. If the object passed to this function is not a
+// PipelineRun, the function will return false.
+func isChainsDoneWithPipelineRun(objectNew client.Object) bool {
 	if newPipelineRun, ok := objectNew.(*tektonv1beta1.PipelineRun); ok {
-		return metadata.HasAnnotationWithValue(newPipelineRun, PipelineRunChainsSignedAnnotation, "true")
+		// If the annotation value is set "true" it means Chains was able to sign. If it is set to
+		// "failed", it means something prevented Chains from signing, e.g. insufficient access. In
+		// either case, we want to proceed processing the PipelineRun instead of waiting
+		// indefinitely for a condition that may never happen, e.g. "failed" -> "true". Let
+		// downstream verification processes, e.g. Enterprise Contract, deal with the Chains
+		// failure.
+		return metadata.HasAnnotation(newPipelineRun, PipelineRunChainsSignedAnnotation)
```

Comment:
just to confirm: if we go by this approach, then we would only be checking if this annotation is present within the PLR even though it: 1) has no value, 2) is set to "true", or 3) is set to "failed".

Is it possible that this annotation could at some point be empty (or has no value) and cause this bug ([RHTAPBUGS-214](https://issues.redhat.com/browse/RHTAPBUGS-214)) again? Another part of me says it won't hit that bug because `Missing pipelinerun attestation` means _absence_ of this annotation.

Reply:
Looking at the Chains code base does, the Chains controller does not set that annotation to an empty value. So this would only happen if something else does that. (But if there's something else playing with that annotation we're already in trouble.)

You may see `Missing pipelinerun attestation` again if Chains does fail to process the PipelineRun, but not because it hasn't had a chance to do so. This is the correct behavior IMO.

Reply:
That makes sense, thanks for the explanation!